frontend/src/lib/ReportNode.ts

```diff
@@ -0,0 +1,509 @@
+  /**
+   * The ID of the report this node was generated in. When a report finishes
+   * running, the app prunes all stale nodes.
+   */
+  readonly reportId: string
+
+  /**
+   * Return the ReportNode for the given index path, or undefined if the path
+   * is invalid.
+   */
+  getIn(path: number[]): ReportNode | undefined
+
+  /**
+   * Return a copy of this node with a new element set at the given index
+   * path. Throws an error if the path is invalid.
+   */
+  setIn(path: number[], node: ReportNode, reportId: string): ReportNode
+
+  /**
+   * Recursively remove children nodes whose reportID is no longer current.
+   * If this node should no longer exist, return undefined.
+   */
+  clearStaleNodes(currentReportId: string): ReportNode | undefined
+
+  /**
+   * Recursively add all Elements contained in the tree to a set.
+   */
+  getElements(elementSet: Set<Element>): void
+}
+
+/**
+ * A leaf ReportNode. Contains a single element to render.
+ */
+export class ElementNode implements ReportNode {
+  public readonly element: Element
+
+  public readonly metadata: ForwardMsgMetadata
+
+  public readonly reportId: string
+
+  /**
+   * A lazily-created immutableJS version of our element.
+   *
+   * This is a temporary! `immutableElement` is currently needed for
+   * dataframe-consuming elements because our dataframe API is
+   * immutableJS-based. It'll go away when we've converted to an ArrowJS-based
+   * dataframe API.
+   *
+   * Because most elements do *not* use the Dataframe API, and therefore
+   * do not need to access `immutableElement`, it is calculated lazily.
+   */
+  private lazyImmutableElement: ImmutableMap<string, any> | undefined
+
+  /** Create a new ElementNode. */
+  public constructor(
+    element: Element,
+    metadata: ForwardMsgMetadata,
+    reportId: string
+  ) {
+    this.element = element
+    this.metadata = metadata
+    this.reportId = reportId
+  }
+
+  public get immutableElement(): ImmutableMap<string, any> {
+    if (this.lazyImmutableElement !== undefined) {
+      return this.lazyImmutableElement
+    }
+
+    const toReturn = toImmutableProto(Element, this.element)
+    this.lazyImmutableElement = toReturn
+    return toReturn
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  public getIn(path: number[]): ReportNode | undefined {
+    return undefined
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  public setIn(
+    path: number[],
+    node: ReportNode,
+    reportId: string
+  ): ReportNode {
+    throw new Error("'setIn' cannot be called on an ElementNode")
+  }
+
+  public clearStaleNodes(currentReportId: string): ElementNode | undefined {
+    return this.reportId === currentReportId ? this : undefined
+  }
+
+  public getElements(elements: Set<Element>): void {
+    elements.add(this.element)
+  }
+
+  public addRows(namedDataSet: NamedDataSet, reportId: string): ElementNode {
+    const newNode = new ElementNode(this.element, this.metadata, reportId)
```

Comment:
Is this different from:

```
const newElement= addRows(
      this.immutableElement,
      toImmutableProto(NamedDataSet, namedDataSet)
    )
return new ElementNode(newElement, this.metadata, reportId)
```
?

Reply:
Yeah - the return value of `addRows` is an `ImmutableMap` (not an `Element` protobuf). So we can't use it to instantiate a new `ElementNode`.